plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java

```diff
@@ -0,0 +1,196 @@
+import org.sonatype.nexus.repository.view.payloads.StringPayload;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+import com.google.common.hash.HashCode;
+import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Reader;
+import org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Writer;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Maven specific implementation of {@link GroupFacetImpl}.
+ *
+ * @since 3.0
+ */
+@Named
+@Facet.Exposed
+public class MavenGroupFacet
+    extends GroupFacetImpl
+{
+  private final Maven2MetadataMerger metadataMerger;
+
+  private MavenFacet mavenFacet;
+
+  @Inject
+  public MavenGroupFacet(final RepositoryManager repositoryManager) {
+    super(repositoryManager);
+    this.metadataMerger = new Maven2MetadataMerger();
+  }
+
+  @Override
+  protected void doConfigure() throws Exception {
+    super.doConfigure();
+    this.mavenFacet = getRepository().facet(MavenFacet.class);
+  }
+
+  /**
+   * Fetches cached metadata if exists, or {@code null}.
+   */
+  @Nullable
+  public Content getCachedMergedMetadata(final MavenPath mavenPath) throws IOException {
+    return mavenFacet.get(mavenPath);
+  }
+
+  /**
+   * Merges and caches and returns the merged metadata. Returns {@code null} if no usable response was in passed in
+   * map.
+   */
+  @Nullable
+  public Content mergeAndCacheMetadata(final MavenPath mavenPath,
+                                       final LinkedHashMap<Repository, Response> responses) throws IOException
+  {
+    checkArgument(mavenFacet.getMavenPathParser().isRepositoryMetadata(mavenPath),
+        "Only metadata can be merged and cached: %s", mavenPath);
+    checkArgument(!mavenPath.isSubordinate(), "Only metadata XML can be merged and cached: %s", mavenPath);
+    final LinkedHashMap<Repository, Content> metadataContents = Maps.newLinkedHashMap();
+    for (Map.Entry<Repository, Response> entry : responses.entrySet()) {
+      if (entry.getValue().getStatus().getCode() == HttpStatus.OK) { // this MUST be 200 OK and no other 2xx code
+        final Response response = entry.getValue();
+        if (response instanceof PayloadResponse && ((PayloadResponse) response).getPayload() instanceof Content) {
+          metadataContents.put(entry.getKey(), (Content) ((PayloadResponse) response).getPayload());
+        }
+      }
+    }
+
+    if (metadataContents.isEmpty()) {
+      return null;
+    }
+    return doMergeAndCacheMetadata(mavenPath, metadataContents);
+  }
+
+  private Content doMergeAndCacheMetadata(final MavenPath mavenPath,
```

Comment:
I think a 50-line method with the word 'and' in its name is a clear sign that it's doing too much in a single method. :P